File: android/app/src/main/java/com/francescooddo/buddygrammar/core/KeyboardLatencyRecorder.kt
package com.francescooddo.buddygrammar.core

import java.util.EnumMap
import kotlin.math.ceil

/** Content-free latency categories collected by the keyboard process. */
enum class KeyboardLatencyMetric {
    KEY_DOWN_TO_FEEDBACK,
    KEY_DOWN_TO_COMMIT,
    SWIPE_DECODE,
}

/** Injectable monotonic clock for deterministic tests. */
fun interface KeyboardLatencyClock {
    fun nowNanoseconds(): Long
}

object SystemKeyboardLatencyClock : KeyboardLatencyClock {
    override fun nowNanoseconds(): Long = System.nanoTime()
}

/** Opaque in-memory event pairing token; it cannot carry input content. */
class KeyboardLatencyToken internal constructor(
    internal val identifier: Long,
    internal val metric: KeyboardLatencyMetric,
)

data class KeyboardLatencySummary(
    /** Valid samples seen since this process-local recorder was created. */
    val count: Long,
    /** Recent samples represented by the percentile window. */
    val windowCount: Int,
    val p50Milliseconds: Double?,
    val p95Milliseconds: Double?,
    val p99Milliseconds: Double?,
)

/**
 * Aggregate-only diagnostics. There are no raw durations, tokens, ordered
 * events, characters, text, document context, or swipe coordinates here.
 */
data class KeyboardLatencySnapshot(
    val keyDownToFeedback: KeyboardLatencySummary,
    val keyDownToCommit: KeyboardLatencySummary,
    val swipeDecode: KeyboardLatencySummary,
    val droppedSampleCount: Long,
    val duplicateEventCount: Long,
    val lostEventCount: Long,
    val inFlightEventCount: Int,
) {
    fun summary(metric: KeyboardLatencyMetric): KeyboardLatencySummary = when (metric) {
        KeyboardLatencyMetric.KEY_DOWN_TO_FEEDBACK -> keyDownToFeedback
        KeyboardLatencyMetric.KEY_DOWN_TO_COMMIT -> keyDownToCommit
        KeyboardLatencyMetric.SWIPE_DECODE -> swipeDecode
    }
}

/**
 * Bounded, content-free, process-memory-only recorder for keyboard hot paths.
 *
 * Recording never logs, persists, performs network work, or sorts. The small
 * recent-duration windows are sorted only on an explicit snapshot request.
 */
class KeyboardLatencyRecorder(
    capacityPerMetric: Int = DEFAULT_CAPACITY_PER_METRIC,
    maximumInFlightEvents: Int = DEFAULT_MAXIMUM_IN_FLIGHT_EVENTS,
    recentTokenCapacity: Int = DEFAULT_RECENT_TOKEN_CAPACITY,
    maximumDurationMilliseconds: Double = DEFAULT_MAXIMUM_DURATION_MILLISECONDS,
    private val clock: KeyboardLatencyClock = SystemKeyboardLatencyClock,
) {
    private data class ActiveEvent(
        val metric: KeyboardLatencyMetric,
        val startedAtNanoseconds: Long,
    )

    private class SampleWindow(private val capacity: Int) {
        private val values = LongArray(capacity)
        private var size = 0
        private var nextIndex = 0
        var totalCount: Long = 0
            private set

        /** Returns true when a previous sample was evicted. */
        fun append(value: Long): Boolean {
            totalCount = incremented(totalCount)
            if (size < capacity) {
                values[size] = value
                size += 1
                return false
            }
            values[nextIndex] = value
            nextIndex = (nextIndex + 1) % capacity
            return true
        }

        fun summary(): KeyboardLatencySummary {
            val sorted = values.copyOf(size).apply(LongArray::sort)
            return KeyboardLatencySummary(
                count = totalCount,
                windowCount = size,
                p50Milliseconds = percentile(0.50, sorted),
                p95Milliseconds = percentile(0.95, sorted),
                p99Milliseconds = percentile(0.99, sorted),
            )
        }

        private fun percentile(percentile: Double, sorted: LongArray): Double? {
            if (sorted.isEmpty()) return null
            val rank = ceil(percentile * sorted.size).toInt()
            val index = (rank - 1).coerceIn(0, sorted.lastIndex)
            return sorted[index] / 1_000_000.0
        }
    }

    private val capacityPerMetric = capacityPerMetric.coerceIn(1, HARD_MAXIMUM_CAPACITY_PER_METRIC)
    private val maximumInFlightEvents = maximumInFlightEvents.coerceIn(
        1,
        HARD_MAXIMUM_IN_FLIGHT_EVENTS,
    )
    private val recentTokenCapacity = recentTokenCapacity.coerceIn(
        1,
        HARD_MAXIMUM_RECENT_TOKEN_CAPACITY,
    )
    private val maximumDurationNanoseconds = (maximumDurationMilliseconds
        .takeIf(Double::isFinite) ?: DEFAULT_MAXIMUM_DURATION_MILLISECONDS)
        .coerceIn(
        1.0,
        DEFAULT_MAXIMUM_DURATION_MILLISECONDS,
    ).times(1_000_000.0).toLong()

    private var nextIdentifier = 0L
    private val activeEvents = mutableMapOf<Long, ActiveEvent>()
    private val recentTerminalIdentifiers = ArrayDeque<Long>()
    private val recentTerminalIdentifierSet = mutableSetOf<Long>()
    private val windows = EnumMap<KeyboardLatencyMetric, SampleWindow>(
        KeyboardLatencyMetric::class.java,
    ).apply {
        KeyboardLatencyMetric.entries.forEach { metric ->
            put(metric, SampleWindow(this@KeyboardLatencyRecorder.capacityPerMetric))
        }
    }
    private var droppedSampleCount = 0L
    private var duplicateEventCount = 0L
    private var lostEventCount = 0L

    @Synchronized
    fun begin(metric: KeyboardLatencyMetric): KeyboardLatencyToken {
        if (activeEvents.size >= maximumInFlightEvents) {
            activeEvents.keys.minOrNull()?.let { oldestIdentifier ->
                activeEvents.remove(oldestIdentifier)
                droppedSampleCount = incremented(droppedSampleCount)
                lostEventCount = incremented(lostEventCount)
                rememberTerminal(oldestIdentifier)
            }
        }

        nextIdentifier = if (nextIdentifier == Long.MAX_VALUE) 1L else nextIdentifier + 1L
        val token = KeyboardLatencyToken(nextIdentifier, metric)
        activeEvents[token.identifier] = ActiveEvent(metric, clock.nowNanoseconds())
        return token
    }

    @Synchronized
    fun finish(token: KeyboardLatencyToken) {
        val event = activeEvents.remove(token.identifier)
        if (event == null) {
            recordMissingTerminal(token.identifier)
            return
        }
        if (event.metric != token.metric) {
            lostEventCount = incremented(lostEventCount)
            droppedSampleCount = incremented(droppedSampleCount)
            rememberTerminal(token.identifier)
            return
        }

        val duration = clock.nowNanoseconds() - event.startedAtNanoseconds
        if (duration < 0L || duration > maximumDurationNanoseconds) {
            droppedSampleCount = incremented(droppedSampleCount)
            rememberTerminal(token.identifier)
            return
        }

        if (windows.getValue(event.metric).append(duration)) {
            droppedSampleCount = incremented(droppedSampleCount)
        }
        rememberTerminal(token.identifier)
    }

    /** Cancelling an unfinished event is visible as one dropped sample. */
    @Synchronized
    fun cancel(token: KeyboardLatencyToken) {
        if (activeEvents.remove(token.identifier) == null) {
            recordMissingTerminal(token.identifier)
            return
        }
        droppedSampleCount = incremented(droppedSampleCount)
        rememberTerminal(token.identifier)
    }

    fun <T> measure(metric: KeyboardLatencyMetric, operation: () -> T): T {
        val token = begin(metric)
        return try {
            operation()
        } finally {
            finish(token)
        }
    }

    @Synchronized
    fun snapshot(): KeyboardLatencySnapshot = KeyboardLatencySnapshot(
        keyDownToFeedback = windows.getValue(KeyboardLatencyMetric.KEY_DOWN_TO_FEEDBACK).summary(),
        keyDownToCommit = windows.getValue(KeyboardLatencyMetric.KEY_DOWN_TO_COMMIT).summary(),
        swipeDecode = windows.getValue(KeyboardLatencyMetric.SWIPE_DECODE).summary(),
        droppedSampleCount = droppedSampleCount,
        duplicateEventCount = duplicateEventCount,
        lostEventCount = lostEventCount,
        inFlightEventCount = activeEvents.size,
    )

    private fun recordMissingTerminal(identifier: Long) {
        if (identifier in recentTerminalIdentifierSet) {
            duplicateEventCount = incremented(duplicateEventCount)
        } else {
            lostEventCount = incremented(lostEventCount)
        }
    }

    private fun rememberTerminal(identifier: Long) {
        if (recentTerminalIdentifierSet.add(identifier)) {
            recentTerminalIdentifiers.addLast(identifier)
        }
        while (recentTerminalIdentifiers.size > recentTokenCapacity) {
            recentTerminalIdentifierSet.remove(recentTerminalIdentifiers.removeFirst())
        }
    }

    companion object {
        const val DEFAULT_CAPACITY_PER_METRIC = 256
        const val HARD_MAXIMUM_CAPACITY_PER_METRIC = 512
        const val DEFAULT_MAXIMUM_IN_FLIGHT_EVENTS = 32
        const val HARD_MAXIMUM_IN_FLIGHT_EVENTS = 64
        const val DEFAULT_RECENT_TOKEN_CAPACITY = 64
        const val HARD_MAXIMUM_RECENT_TOKEN_CAPACITY = 128
        const val DEFAULT_MAXIMUM_DURATION_MILLISECONDS = 60_000.0

        /** Shared, process-local recorder used by production keyboard seams. */
        val production = KeyboardLatencyRecorder()

        private fun incremented(value: Long): Long = if (value == Long.MAX_VALUE) value else value + 1
    }
}

File: BuddyGrammarKit/Sources/BuddyGrammarKit/KeyboardLatencyRecorder.swift
import Dispatch
import Foundation

/// The only latency categories collected by the keyboard. No category carries
/// key labels, committed text, document context, or gesture geometry.
public enum KeyboardLatencyMetric: CaseIterable, Equatable, Hashable, Sendable {
    case keyDownToFeedback
    case keyDownToCommit
    case swipeDecode
}

/// Injectable monotonic time source used by ``KeyboardLatencyRecorder``.
public protocol KeyboardLatencyClock: Sendable {
    func nowNanoseconds() -> UInt64
}

public struct SystemKeyboardLatencyClock: KeyboardLatencyClock {
    public init() {}

    public func nowNanoseconds() -> UInt64 {
        DispatchTime.now().uptimeNanoseconds
    }
}

/// An opaque pairing token. It identifies only an in-memory timing event and
/// cannot hold input content.
public struct KeyboardLatencyToken: Equatable, Hashable, Sendable {
    fileprivate let identifier: UInt64
    fileprivate let metric: KeyboardLatencyMetric
}

public struct KeyboardLatencySummary: Equatable, Sendable {
    /// Number of valid samples observed since this in-memory recorder started.
    public let count: UInt64
    /// Number of recent samples represented by the percentile window.
    public let windowCount: Int
    public let p50Milliseconds: Double?
    public let p95Milliseconds: Double?
    public let p99Milliseconds: Double?
}

/// Aggregate-only diagnostics. The snapshot contains no raw durations, event
/// identifiers, ordered events, characters, text, or swipe coordinates.
public struct KeyboardLatencySnapshot: Equatable, Sendable {
    public let keyDownToFeedback: KeyboardLatencySummary
    public let keyDownToCommit: KeyboardLatencySummary
    public let swipeDecode: KeyboardLatencySummary
    public let droppedSampleCount: UInt64
    public let duplicateEventCount: UInt64
    public let lostEventCount: UInt64
    public let inFlightEventCount: Int

    public func summary(for metric: KeyboardLatencyMetric) -> KeyboardLatencySummary {
        switch metric {
        case .keyDownToFeedback: keyDownToFeedback
        case .keyDownToCommit: keyDownToCommit
        case .swipeDecode: swipeDecode
        }
    }
}

/// A content-free, process-memory-only latency recorder for keyboard hot paths.
///
/// Recording performs bounded dictionary/array mutation under a short lock. It
/// never logs, persists, performs network work, or sorts on a hot path. Sorting
/// happens only when an aggregate snapshot is explicitly requested.
public final class KeyboardLatencyRecorder: @unchecked Sendable {
    public static let defaultCapacityPerMetric = 256
    public static let hardMaximumCapacityPerMetric = 512
    public static let defaultMaximumInFlightEvents = 32
    public static let hardMaximumInFlightEvents = 64
    public static let defaultRecentTokenCapacity = 64
    public static let hardMaximumRecentTokenCapacity = 128
    public static let defaultMaximumDurationMilliseconds: Double = 60_000

    /// Shared process-local recorder used by production keyboard seams.
    public static let production = KeyboardLatencyRecorder()

    private struct ActiveEvent {
        let metric: KeyboardLatencyMetric
        let startedAtNanoseconds: UInt64
    }

    /// Reference semantics are intentional: the dictionary never copy-on-writes
    /// a duration array when a sample is appended under the recorder lock.
    private final class SampleWindow {
        var values: [UInt64]
        var nextIndex = 0
        var totalCount: UInt64 = 0

        init(capacity: Int) {
            values = []
            values.reserveCapacity(capacity)
        }

        func append(_ value: UInt64, capacity: Int) -> Bool {
            if totalCount < .max { totalCount += 1 }
            if values.count < capacity {
                values.append(value)
                return false
            }
            values[nextIndex] = value
            nextIndex = (nextIndex + 1) % capacity
            return true
        }

        func summary() -> KeyboardLatencySummary {
            let sorted = values.sorted()
            return KeyboardLatencySummary(
                count: totalCount,
                windowCount: sorted.count,
                p50Milliseconds: percentile(0.50, in: sorted),
                p95Milliseconds: percentile(0.95, in: sorted),
                p99Milliseconds: percentile(0.99, in: sorted)
            )
        }

        private func percentile(_ percentile: Double, in sorted: [UInt64]) -> Double? {
            guard !sorted.isEmpty else { return nil }
            let rank = Int(ceil(percentile * Double(sorted.count)))
            let index = min(sorted.count - 1, max(0, rank - 1))
            return Double(sorted[index]) / 1_000_000
        }
    }

    private let lock = NSLock()
    private let capacityPerMetric: Int
    private let maximumInFlightEvents: Int
    private let recentTokenCapacity: Int
    private let maximumDurationNanoseconds: UInt64
    private let clock: any KeyboardLatencyClock

    private var nextIdentifier: UInt64 = 0
    private var activeEvents: [UInt64: ActiveEvent] = [:]
    private var recentTerminalIdentifiers: [UInt64] = []
    private var recentTerminalNextIndex = 0
    private var recentTerminalIdentifierSet: Set<UInt64> = []
    private let windows: [KeyboardLatencyMetric: SampleWindow]
    private var droppedSampleCount: UInt64 = 0
    private var duplicateEventCount: UInt64 = 0
    private var lostEventCount: UInt64 = 0

    public init(
        capacityPerMetric: Int = defaultCapacityPerMetric,
        maximumInFlightEvents: Int = defaultMaximumInFlightEvents,
        recentTokenCapacity: Int = defaultRecentTokenCapacity,
        maximumDurationMilliseconds: Double = defaultMaximumDurationMilliseconds,
        clock: any KeyboardLatencyClock = SystemKeyboardLatencyClock()
    ) {
        let resolvedCapacityPerMetric = min(
            Self.hardMaximumCapacityPerMetric,
            max(1, capacityPerMetric)
        )
        let resolvedMaximumInFlightEvents = min(
            Self.hardMaximumInFlightEvents,
            max(1, maximumInFlightEvents)
        )
        let resolvedRecentTokenCapacity = min(
            Self.hardMaximumRecentTokenCapacity,
            max(1, recentTokenCapacity)
        )
        self.capacityPerMetric = resolvedCapacityPerMetric
        self.maximumInFlightEvents = resolvedMaximumInFlightEvents
        self.recentTokenCapacity = resolvedRecentTokenCapacity
        let finiteDuration = maximumDurationMilliseconds.isFinite
            ? maximumDurationMilliseconds
            : Self.defaultMaximumDurationMilliseconds
        let clampedDuration = min(
            Self.defaultMaximumDurationMilliseconds,
            max(1, finiteDuration)
        )
        self.maximumDurationNanoseconds = UInt64(clampedDuration * 1_000_000)
        self.clock = clock
        self.windows = Dictionary(
            uniqueKeysWithValues: KeyboardLatencyMetric.allCases.map {
                ($0, SampleWindow(capacity: resolvedCapacityPerMetric))
            }
        )
        self.activeEvents.reserveCapacity(resolvedMaximumInFlightEvents)
        self.recentTerminalIdentifiers.reserveCapacity(resolvedRecentTokenCapacity)
        self.recentTerminalIdentifierSet.reserveCapacity(resolvedRecentTokenCapacity)
    }

    public func begin(_ metric: KeyboardLatencyMetric) -> KeyboardLatencyToken {
        lock.lock()
        defer { lock.unlock() }

        if activeEvents.count >= maximumInFlightEvents,
           let oldestIdentifier = activeEvents.keys.min() {
            activeEvents.removeValue(forKey: oldestIdentifier)
            increment(&droppedSampleCount)
            increment(&lostEventCount)
            rememberTerminal(oldestIdentifier)
        }

        nextIdentifier &+= 1
        if nextIdentifier == 0 { nextIdentifier = 1 }
        let token = KeyboardLatencyToken(identifier: nextIdentifier, metric: metric)
        activeEvents[token.identifier] = ActiveEvent(
            metric: metric,
            startedAtNanoseconds: clock.nowNanoseconds()
        )
        return token
    }

    public func finish(_ token: KeyboardLatencyToken) {
        lock.lock()
        defer { lock.unlock() }

        guard let event = activeEvents.removeValue(forKey: token.identifier) else {
            recordMissingTerminal(token.identifier)
            return
        }
        guard event.metric == token.metric else {
            increment(&lostEventCount)
            increment(&droppedSampleCount)
            rememberTerminal(token.identifier)
            return
        }

        let finishedAt = clock.nowNanoseconds()
        guard finishedAt >= event.startedAtNanoseconds else {
            increment(&droppedSampleCount)
            rememberTerminal(token.identifier)
            return
        }
        let duration = finishedAt - event.startedAtNanoseconds
        guard duration <= maximumDurationNanoseconds else {
            increment(&droppedSampleCount)
            rememberTerminal(token.identifier)
            return
        }

        guard let window = windows[event.metric] else {
            increment(&lostEventCount)
            increment(&droppedSampleCount)
            rememberTerminal(token.identifier)
            return
        }
        if window.append(duration, capacity: capacityPerMetric) {
            increment(&droppedSampleCount)
        }
        rememberTerminal(token.identifier)
    }

    /// Cancels an incomplete measurement. Cancellation discards no content;
    /// it increments the dropped counter so lifecycle gaps remain visible.
    public func cancel(_ token: KeyboardLatencyToken) {
        lock.lock()
        defer { lock.unlock() }

        guard activeEvents.removeValue(forKey: token.identifier) != nil else {
            recordMissingTerminal(token.identifier)
            return
        }
        increment(&droppedSampleCount)
        rememberTerminal(token.identifier)
    }

    @discardableResult
    public func measure<T>(
        _ metric: KeyboardLatencyMetric,
        operation: () throws -> T
    ) rethrows -> T {
        let token = begin(metric)
        defer { finish(token) }
        return try operation()
    }

    public func snapshot() -> KeyboardLatencySnapshot {
        lock.lock()
        defer { lock.unlock() }

        return KeyboardLatencySnapshot(
            keyDownToFeedback: windows[.keyDownToFeedback]!.summary(),
            keyDownToCommit: windows[.keyDownToCommit]!.summary(),
            swipeDecode: windows[.swipeDecode]!.summary(),
            droppedSampleCount: droppedSampleCount,
            duplicateEventCount: duplicateEventCount,
            lostEventCount: lostEventCount,
            inFlightEventCount: activeEvents.count
        )
    }

    private func recordMissingTerminal(_ identifier: UInt64) {
        if recentTerminalIdentifierSet.contains(identifier) {
            increment(&duplicateEventCount)
        } else {
            increment(&lostEventCount)
        }
    }

    private func rememberTerminal(_ identifier: UInt64) {
        guard recentTerminalIdentifierSet.insert(identifier).inserted else { return }
        if recentTerminalIdentifiers.count < recentTokenCapacity {
            recentTerminalIdentifiers.append(identifier)
            return
        }
        let removed = recentTerminalIdentifiers[recentTerminalNextIndex]
        recentTerminalIdentifierSet.remove(removed)
        recentTerminalIdentifiers[recentTerminalNextIndex] = identifier
        recentTerminalNextIndex = (recentTerminalNextIndex + 1) % recentTokenCapacity
    }

    private func increment(_ value: inout UInt64) {
        if value < .max { value += 1 }
    }
}
